File: DatabindingWithRecyclerView/app/src/main/java/com/enpassio/databindingwithrecyclerview/ui/ProductAdapter.java
package com.enpassio.databindingwithrecyclerview.ui;

import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.databinding.DataBindingUtil;
import androidx.recyclerview.widget.RecyclerView;

import com.enpassio.databindingwithrecyclerview.R;
import com.enpassio.databindingwithrecyclerview.databinding.ItemProductBinding;
import com.enpassio.databindingwithrecyclerview.model.Product;

import java.util.List;

public class ProductAdapter extends RecyclerView.Adapter<ProductAdapter.ProductViewHolder> {

    private final List<Product> mProductList;
    private final ProductItemClickListener mListener;

    ProductAdapter(@NonNull List<Product> productList, ProductItemClickListener listener) {
        mProductList =productList;
        mListener = listener;
    }

    @NonNull
    @Override
    public ProductViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        ItemProductBinding binding = DataBindingUtil
                .inflate(LayoutInflater.from(parent.getContext()), R.layout.item_product,
                        parent, false);

        return new ProductViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull ProductViewHolder holder, int position) {
        holder.bind(mProductList.get(position), mListener);
    }

    @Override
    public int getItemCount() {
        return mProductList.size();
    }

    class ProductViewHolder extends RecyclerView.ViewHolder{

        final ItemProductBinding binding;

        ProductViewHolder(ItemProductBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }

        void bind(Product currentProduct, ProductItemClickListener clickListener){
            //For each item, corresponding product object is passed to the binding
            binding.setProduct(currentProduct);
            binding.setProductItemClick(clickListener);
            //This is to force bindings to execute right away
            binding.executePendingBindings();
        }
    }

    public interface ProductItemClickListener {
        void onProductItemClicked(Product product);
    }
}

File: DatabindingWithRecyclerViewKotlin/app/src/main/java/com/enpassio/databindingwithrecyclerviewkotlin/ui/ProductAdapter.kt
package com.enpassio.databindingwithrecyclerviewkotlin.ui

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.recyclerview.widget.RecyclerView
import com.enpassio.databindingwithrecyclerviewkotlin.R
import com.enpassio.databindingwithrecyclerviewkotlin.databinding.ItemProductBinding
import com.enpassio.databindingwithrecyclerviewkotlin.model.Product


/**
 * Created by Greta Grigutė on 2018-12-23.
 */
class ProductAdapter internal constructor(
    private val mProductList: List<Product>,
    private val mListener: ProductItemClickListener
) : RecyclerView.Adapter<ProductAdapter.ProductViewHolder>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ProductViewHolder = ProductViewHolder.from(parent)

    override fun onBindViewHolder(holder: ProductViewHolder, position: Int) = holder.bind(mProductList[position], mListener)

    override fun getItemCount(): Int = mProductList.size

    class ProductViewHolder(val binding: ItemProductBinding) : RecyclerView.ViewHolder(binding.root){

        fun bind(currentProduct: Product, listener : ProductItemClickListener){
            binding.product = currentProduct
            binding.productItemClick = listener
            binding.executePendingBindings()
        }

        companion object {
            fun from(parent: ViewGroup): ProductViewHolder {
                val layoutInflater = LayoutInflater.from(parent.context)
                val binding: ItemProductBinding = DataBindingUtil
                    .inflate(layoutInflater, R.layout.item_product,
                        parent, false)
                return ProductViewHolder(binding)
            }
        }
    }

    interface ProductItemClickListener {
        fun onProductItemClicked(product: Product)
    }
}